services/wallet/history/service.go

```diff
@@ -0,0 +1,235 @@
+	EventBalanceHistoryUpdateFinished          walletevent.EventType = "wallet-balance-history-update-finished"
+	EventBalanceHistoryUpdateFinishedWithError walletevent.EventType = "wallet-balance-history-update-finished-with-error"
+
+	balanceHistoryUpdateInterval = 12 * time.Hour
+)
+
+type Service struct {
+	balance        *Balance
+	db             *sql.DB
+	eventFeed      *event.Feed
+	rpcClient      *statusrpc.Client
+	networkManager *network.Manager
+	tokenManager   *token.Manager
+	serviceContext context.Context
+	cancelFn       context.CancelFunc
+}
+
+func NewService(db *sql.DB, eventFeed *event.Feed, rpcClient *statusrpc.Client, tokenManager *token.Manager) *Service {
+	return &Service{
+		balance:        NewBalance(NewBalanceDB(db)),
+		db:             db,
+		eventFeed:      eventFeed,
+		rpcClient:      rpcClient,
+		networkManager: rpcClient.NetworkManager,
+		tokenManager:   tokenManager,
+	}
+}
+
+func (s *Service) Stop() {
+	s.cancelFn()
+}
+
+func (s *Service) triggerEvent(eventType walletevent.EventType, account statustypes.Address, message string) {
+	s.eventFeed.Send(walletevent.Event{
+		Type: eventType,
+		Accounts: []common.Address{
+			common.Address(account),
+		},
+		Message: message,
+	})
+}
+
+func (s *Service) StartBalanceHistory() {
+	go func() {
+		s.serviceContext, s.cancelFn = context.WithCancel(context.Background())
+		timer := time.NewTimer(balanceHistoryUpdateInterval)
+
+		update := func() {
+			err := s.updateBalanceHistoryForAllEnabledNetworks(s.serviceContext)
+			if s.serviceContext.Err() != nil {
+				s.triggerEvent(EventBalanceHistoryUpdateFinished, statustypes.Address{}, "Service canceled")
+				timer.Stop()
+				return
+			}
+			if err != nil {
+				s.triggerEvent(EventBalanceHistoryUpdateFinishedWithError, statustypes.Address{}, err.Error())
+			}
+		}
+
+		update()
+
+		for range timer.C {
+			timer.Reset(balanceHistoryUpdateInterval)
+
+			update()
+		}
+	}()
+}
+
+// Native token implementation of BlockInfoSource interface
+type chainClientSource struct {
+	chainClient *chain.Client
+	currency    string
+}
+
+func (src *chainClientSource) HeaderByNumber(ctx context.Context, number *big.Int) (*types.Header, error) {
+	return src.chainClient.HeaderByNumber(ctx, number)
+}
+
+func (src *chainClientSource) BalanceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (*big.Int, error) {
+	return src.chainClient.BalanceAt(ctx, account, blockNumber)
+}
+
+func (src *chainClientSource) ChainID() uint64 {
+	return src.chainClient.ChainID
+}
+
+func (src *chainClientSource) Currency() string {
+	return src.currency
+}
+
+func (src *chainClientSource) TimeNow() int64 {
+	return time.Now().UTC().Unix()
+}
+
+type tokenChainClientSource struct {
+	chainClientSource
+	TokenManager   *token.Manager
+	NetworkManager *network.Manager
+}
+
+func (src *tokenChainClientSource) BalanceAt(ctx context.Context, account common.Address, blockNumber *big.Int) (*big.Int, error) {
```

Comment:
When fetching balances, there is a contract to fetch balances for multiple token, do u know if that works with block number?

Reply:
this contract works for ERC20 and for native chain balance. But here clearly the interest is for erc20